File: app/src/main/java/com/diu/yk_games/line2box/presentation/online/ChatFragmentFriendly.java
package com.diu.yk_games.line2box.presentation.online;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.AlertDialog;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.SharedPreferences;
import android.graphics.drawable.ColorDrawable;
import android.media.MediaPlayer;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.fragment.app.Fragment;

import android.os.Handler;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import com.diu.yk_games.line2box.presentation.MsgListAdapter;
import com.diu.yk_games.line2box.R;
import com.diu.yk_games.line2box.model.GameProfile;
import com.diu.yk_games.line2box.model.MsgStore;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;


import pl.droidsonroids.gif.GifImageView;


public class ChatFragmentFriendly extends Fragment {
    ArrayList<MsgStore> msList;
    FirebaseDatabase database = FirebaseDatabase.getInstance();
    DatabaseReference myRef;
    EditText chatBoxFriendly;
    String tempMsg=null, lastMsg=null;
    static String key=null;
    Context context;
    Activity activity;
    String playerId;
    SharedPreferences sharedPref;

    public static ChatFragmentFriendly newInstance(String key, String playerId)
    {
        ChatFragmentFriendly fragment = new ChatFragmentFriendly();
        Bundle args = new Bundle();
        args.putString("key", key);
        args.putString("playerId", playerId);
        fragment.setArguments(args);
        return fragment;
    }

    public ChatFragmentFriendly() {
        // Required empty public constructor
    }


    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            key = getArguments().getString("key");
            playerId = getArguments().getString("playerId");
        }
        myRef = database.getReference("MultiPlayer").child(key).child("friendlyChat");

    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View v= inflater.inflate(R.layout.fragment_chat_friendly, container, false);
        chatBoxFriendly = v.findViewById(R.id.chatBoxFriendly);
        chatBoxFriendly.isFocusableInTouchMode();
        chatBoxFriendly.requestFocus();

        context=requireContext();
        activity=requireActivity();
        sharedPref = context.getSharedPreferences(
                getString(R.string.preference_file_key), Context.MODE_PRIVATE);
        GameProfile.setPreferences(sharedPref);

        msList = new ArrayList<>();

        MediaPlayer mp = MediaPlayer.create(context, R.raw.pop);
        DrawerLayout mDrawerLayout = activity.findViewById(R.id.drawer_layout);

        myRef.addChildEventListener(new ChildEventListener()
        {
            @Override
            public void onChildAdded(@NonNull DataSnapshot dataSnapshot, String s)
            {
                if(dataSnapshot.exists())
                {
                    MsgStore ms = dataSnapshot.getValue(MsgStore.class);
                    assert ms != null;
                    msList.add(ms);
                    lastMsg=ms.msgData;
                    if(getActivity()!=null)
                    {
                        try{
                            MsgListAdapter adapter=new MsgListAdapter(getActivity(),msList);  //
                            ListView list = v.findViewById(R.id.showMsgList);
                            list.setAdapter(adapter);
                            list.setOnItemClickListener(new AdapterView.OnItemClickListener() {
                                @Override
                                public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                                    MsgStore msgData = (MsgStore) list.getItemAtPosition(position);
                                    //prestationEco str = (prestationEco)o; //As you are using Default String Adapter

                                    if(!sharedPref.getBoolean("muted", false))
                                    {
                                        MediaPlayer mediaPlayer = MediaPlayer.create(context, R.raw.btn_click_ef);
                                        mediaPlayer.start();
                                        mediaPlayer.setOnCompletionListener(MediaPlayer::release);
                                    }
                                    if(!msgData.playerId.equals(""))
                                    {
                                        Toast.makeText(context,"Long Press To Copy Text/ID",Toast.LENGTH_SHORT).show();
                                        FirebaseFirestore db = FirebaseFirestore.getInstance();
                                        db.collection("gamerProfile").document(msgData.playerId)
                                                .get().addOnSuccessListener(new OnSuccessListener<DocumentSnapshot>() {
                                                    @SuppressLint("SetTextI18n")
                                                    @Override
                                                    public void onSuccess(DocumentSnapshot documentSnapshot) {
                                                        if(documentSnapshot.exists())
                                                        {
                                                            GameProfile server2device = documentSnapshot.toObject(GameProfile.class);
                                                            AlertDialog.Builder builder = new AlertDialog.Builder(context);
                                                            View v = LayoutInflater.from(context).inflate(
                                                                    R.layout.dialog_layout_profile, parent.findViewById(R.id.profileLayoutDialog)
                                                            );
                                                            builder.setView(v);
                                                            LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(
                                                                    LinearLayout.LayoutParams.WRAP_CONTENT,
                                                                    LinearLayout.LayoutParams.WRAP_CONTENT
                                                            );
                                                            params.setMargins(60, 150, 60, 0);
                                                            v.findViewById(R.id.linearLayoutFrame).setLayoutParams(params);
                                                            //v.findViewById(R.id.linearLayoutFrame).setPadding(20,0,20,0);
                                                            assert server2device != null;
                                                            ((TextView) v.findViewById(R.id.countryTxt)).setText(server2device.countryNm+" "+server2device.countryEmoji);
                                                            ((TextView) v.findViewById(R.id.lvlTxt)).setText(""+server2device.lvl);
                                                            ((TextView) v.findViewById(R.id.coinHave)).setText(""+server2device.coin);
                                                            ((TextView) v.findViewById(R.id.matchPlayedTxt)).setText(""+server2device.matchPlayed);
                                                            ((TextView) v.findViewById(R.id.matchWonTxt)).setText(""+server2device.matchWinMulti);
                                                            EditText nmEditText = v.findViewById(R.id.nmTxt);
                                                            nmEditText.setEnabled(false);
                                                            nmEditText.setText(server2device.nm);
                                                            //nmEditText.setVisibility(View.GONE);
                                                            ((TextView) v.findViewById(R.id.profileTitle)).setTextSize(28);

                                                            v.findViewById(R.id.profileShapeLayout).setVisibility(View.GONE);
                                                            v.findViewById(R.id.nmEditBtn).setVisibility(View.GONE);
                                                            v.findViewById(R.id.nmLTxt).setVisibility(View.GONE);
                                                            v.findViewById(R.id.themeBox).setVisibility(View.GONE);
                                                            v.findViewById(R.id.countryLTxt).setVisibility(View.GONE);
                                                            v.findViewById(R.id.buttonSaveInfo).setVisibility(View.GONE);
                                                            final AlertDialog alertDialog = builder.create();

                                                            if (alertDialog.getWindow() != null) {
                                                                alertDialog.getWindow().setBackgroundDrawable(new ColorDrawable(0));
                                                            }
                                                            try {alertDialog.show();}
                                                            catch (NullPointerException npe) {npe.printStackTrace();}
                                                        }
                                                    }
                                                });
                                    }
                                    else
                                        Toast.makeText(context,"Older messages don't have profile info.",Toast.LENGTH_SHORT).show();

                                }
                            });
                            list.setOnItemLongClickListener(new AdapterView.OnItemLongClickListener() {
                                @Override
                                public boolean onItemLongClick(AdapterView<?> parent, View view, int position, long id)
                                {
                                    Toast.makeText(context, "Text/ID copied", Toast.LENGTH_SHORT).show();
                                    ClipboardManager clipboard = (ClipboardManager) context.getSystemService(Context.CLIPBOARD_SERVICE);
                                    ClipData clip = ClipData.newPlainText("msgData", ms.msgData);
                                    clipboard.setPrimaryClip(clip);
                                    return true;
                                }
                            });
                            getActivity().findViewById(R.id.newMsgBoltu).setVisibility(View.VISIBLE);
                        }
                        catch (NullPointerException npe) {npe.printStackTrace();}


                    }
                }

            }
            @Override public void onChildChanged(@NonNull DataSnapshot dataSnapshot, @Nullable String s) {}
            @Override public void onChildRemoved(@NonNull DataSnapshot dataSnapshot) {}
            @Override public void onChildMoved(@NonNull DataSnapshot dataSnapshot, @Nullable String s) {}
            @Override public void onCancelled(@NonNull DatabaseError databaseError) {}
        });
        myRef.addValueEventListener(new ValueEventListener(){
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                if(snapshot.exists())
                {
                    if(lastMsg.equals("🤣"))
                        emojiRunner(R.drawable.emoji_haha,R.raw.haha,v);
                    else if(lastMsg.equals("😭"))
                        emojiRunner(R.drawable.emoji_cry,R.raw.cry,v);
                    else if(lastMsg.equals("😱"))
                        emojiRunner(R.drawable.emoji_scream,R.raw.scream,v);
                    else if(lastMsg.equals("😘"))
                        emojiRunner(R.drawable.emoji_kiss,R.raw.kiss,v);
                    else if(lastMsg.equals("🥱"))
                        emojiRunner(R.drawable.emoji_yawn,R.raw.yawn,v);
                    else if(lastMsg.equals(tempMsg))
                    {
                        MediaPlayer mediaPlayer = MediaPlayer.create(context, R.raw.pop);
                        mediaPlayer.start();
                        mediaPlayer.setOnCompletionListener(MediaPlayer::release);
                    }
                    else if(!mDrawerLayout.isDrawerOpen(GravityCompat.START))
                    {
                        mp.start();
                        //mp.setOnCompletionListener(MediaPlayer::release);
                    }

                    tempMsg=" # # 69";
                }
            }
            @Override public void onCancelled(@NonNull DatabaseError error) {}
        });
        ImageButton sendMsg=v.findViewById(R.id.msgSendBtn);
        sendMsg.setOnClickListener(v1->
        {
            sendThisMsg(chatBoxFriendly.getText().toString());
            chatBoxFriendly.setText("");
        });
        v.findViewById(R.id.sendHaha).setOnClickListener(v1-> sendThisMsg("🤣"));
        v.findViewById(R.id.sendCry).setOnClickListener(v1-> sendThisMsg("😭"));
        v.findViewById(R.id.sendKiss).setOnClickListener(v1-> sendThisMsg("😘"));
        v.findViewById(R.id.sendScream).setOnClickListener(v1-> sendThisMsg("😱"));
        v.findViewById(R.id.sendYawn).setOnClickListener(v1-> sendThisMsg("🥱"));

        return v;
    }
    public void emojiRunner(int gif, int sound, View v)
    {
        v.findViewById(R.id.sendHaha).setEnabled(false);
        v.findViewById(R.id.sendCry).setEnabled(false);
        v.findViewById(R.id.sendKiss).setEnabled(false);
        v.findViewById(R.id.sendScream).setEnabled(false);
        v.findViewById(R.id.sendYawn).setEnabled(false);
        if(!sharedPref.getBoolean("muted", false))
        {
            MediaPlayer mediaPlayer = MediaPlayer.create(context, sound);
            mediaPlayer.start();
            mediaPlayer.setOnCompletionListener(MediaPlayer::release);
        }
        ((GifImageView)activity.findViewById(R.id.emojiPlay)).setImageResource(gif);
        activity.findViewById(R.id.emojiPlay).setVisibility(View.VISIBLE);
        ((DrawerLayout)activity.findViewById(R.id.drawer_layout)).closeDrawer(GravityCompat.START);
        activity.findViewById(R.id.newMsgBoltu).setVisibility(View.GONE);
        Handler handler = new Handler();
        handler.postDelayed(() ->
                {
                    activity.findViewById(R.id.emojiPlay).setVisibility(View.GONE);
                    v.findViewById(R.id.sendHaha).setEnabled(true);
                    v.findViewById(R.id.sendCry).setEnabled(true);
                    v.findViewById(R.id.sendKiss).setEnabled(true);
                    v.findViewById(R.id.sendScream).setEnabled(true);
                    v.findViewById(R.id.sendYawn).setEnabled(true);
                }, 2500);

    }
    public void sendThisMsg(String msg)
    {
        GameProfile gp=new GameProfile();
        MsgStore ms =new MsgStore();
        ms.playerId=playerId;
        ms.nmData= gp.nm;
        ms.lvlData= Integer.toString(gp.getLvlByCal());
        DateTimeFormatter dtf = DateTimeFormatter.ofPattern("dd MMM, hh:mm a");
        LocalDateTime now = LocalDateTime.now();
        ms.timeData = dtf.format(now);
        tempMsg=msg;
        ms.msgData=msg;
        String key = myRef.push().getKey();
        assert key != null;
        myRef.child(key).setValue(ms);
    }
}

//    Handler handler = new Handler();
//    handler.postDelayed(() ->
//            {
//
//            }, 600);

File: app/src/main/java/com/diu/yk_games/line2box/presentation/online/ChatFragmentFriendly.kt
package com.diu.yk_games.line2box.presentation.online

import android.annotation.SuppressLint
import android.app.Activity
import android.app.AlertDialog
import android.media.MediaPlayer
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import android.widget.ImageView
import android.widget.LinearLayout
import androidx.core.graphics.drawable.toDrawable
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.lifecycle.lifecycleScope
import com.diu.yk_games.line2box.R
import com.diu.yk_games.line2box.databinding.DialogLayoutProfileBinding
import com.diu.yk_games.line2box.databinding.FragmentChatFriendlyBinding
import com.diu.yk_games.line2box.model.GameProfile
import com.diu.yk_games.line2box.presentation.MainViewModel
import com.diu.yk_games.line2box.presentation.MsgListAdapter
import com.diu.yk_games.line2box.util.collectWithLifecycle
import com.diu.yk_games.line2box.util.gone
import com.diu.yk_games.line2box.util.loadDrawable
import com.diu.yk_games.line2box.util.log
import com.diu.yk_games.line2box.util.pref
import com.diu.yk_games.line2box.util.setBounceClickListener
import com.diu.yk_games.line2box.util.setClipBoardData
import com.diu.yk_games.line2box.util.show
import com.diu.yk_games.line2box.util.toast
import com.google.firebase.Firebase
import com.google.firebase.firestore.firestore
import com.google.firebase.firestore.toObject
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class ChatFragmentFriendly : Fragment() {
    private lateinit var binding: FragmentChatFriendlyBinding
    private val viewModel by activityViewModels<MainViewModel>()
    private lateinit var parentActivity: Activity
    private val msgListAdapter by lazy { MsgListAdapter(viewModel.playerId) }
    private val drawerLayout by lazy { parentActivity.findViewById<DrawerLayout>(R.id.drawer_layout) }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentChatFriendlyBinding.inflate(inflater, container, false)
        parentActivity = requireActivity()
        binding.showMsgList.adapter = msgListAdapter
        return binding.root
    }

    @SuppressLint("SetTextI18n")
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.chatBoxFriendly.requestFocus()
        val mp = MediaPlayer.create(parentActivity, R.raw.pop)

        viewModel.friendsChatList.collectWithLifecycle {
            it.size.log("friendsChatList")
            msgListAdapter.submitList(it) {
                binding.showMsgList.scrollToPosition(0)
            }
            val lastMsg = it.firstOrNull()
            if (lastMsg?.key == lastMsgKey) return@collectWithLifecycle
            when (lastMsg?.msgData) {
                "🤣" -> emojiRunner(R.drawable.emoji_haha, R.raw.haha)
                "😭" -> emojiRunner(R.drawable.emoji_cry, R.raw.cry)
                "😱" -> emojiRunner(R.drawable.emoji_scream, R.raw.scream)
                "😘" -> emojiRunner(R.drawable.emoji_kiss, R.raw.kiss)
                "🥱" -> emojiRunner(R.drawable.emoji_yawn, R.raw.yawn)
                else -> {
                    mp.start()
                    viewModel.setNewMsgBoltVisible(true)
                }
            }
            lastMsg?.key?.let {
                lastMsgKey = lastMsg.key
            }
        }

        msgListAdapter.onClickListener = { msg ->
            //presentationEco str = (presentationEco)o; //As you are using Default String Adapter
            if (!pref.read("muted", false)) {
                val mediaPlayer =
                    MediaPlayer.create(parentActivity, R.raw.btn_click_ef)
                mediaPlayer.start()
                mediaPlayer.setOnCompletionListener(MediaPlayer::release)
            }
            if (msg.playerId.isNotEmpty()) {
                val db = Firebase.firestore
                db.collection("gamerProfile").document(msg.playerId)
                    .get()
                    .addOnSuccessListener { documentSnapshot ->
                        val server2device = documentSnapshot.toObject<GameProfile>()
                        if (server2device != null) {
                            val builder = AlertDialog.Builder(parentActivity)
                            val binding =
                                DialogLayoutProfileBinding.inflate(layoutInflater)
                            builder.setView(binding.root)
                            val params = LinearLayout.LayoutParams(
                                LinearLayout.LayoutParams.WRAP_CONTENT,
                                LinearLayout.LayoutParams.WRAP_CONTENT
                            )
                            params.setMargins(60, 150, 60, 0)
                            binding.apply {
                                linearLayoutFrame.layoutParams = params
                                countryTxt.text =
                                    "${server2device.countryNm} ${server2device.countryEmoji}"
                                lvlTxt.text = server2device.lvl.toString()
                                coinHave.text = server2device.coin.toString()
                                matchPlayedTxt.text = server2device.matchPlayed.toString()
                                matchWonTxt.text = server2device.matchWinMulti.toString()
                                nmTxt.isEnabled = false
                                nmTxt.setText(server2device.nm)
                                profileTitle.textSize = 28f
                                profileShapeLayout.gone()
                                nmEditBtn.gone()
                                nmLTxt.gone()
                                themeBox.gone()
                                countryLTxt.gone()
                                buttonSaveInfo.gone()
                            }
                            val alertDialog = builder.create()
                            alertDialog.window?.setBackgroundDrawable(0.toDrawable())
                            binding.root.setOnClickListener {
                                runCatching { if (alertDialog.isShowing) alertDialog.dismiss() }
                            }
                            runCatching { alertDialog.show() }
                        }
                    }
            } else
                toast("Older messages don't have profile info.")
        }

        msgListAdapter.onLongClickListener = { msg ->
            parentActivity.setClipBoardData(msg.msgData, "Text copied")
            true
        }
        binding.msgSendBtn.setBounceClickListener {
            viewModel.sendMessage2FriendlyChat(binding.chatBoxFriendly.text.toString())?.also {
                binding.chatBoxFriendly.setText("")
            } ?: toast("Write Something..")
        }
        binding.sendHaha.setBounceClickListener { sendEmoji("🤣") }
        binding.sendCry.setBounceClickListener { sendEmoji("😭") }
        binding.sendKiss.setBounceClickListener { sendEmoji("😘") }
        binding.sendScream.setBounceClickListener { sendEmoji("😱") }
        binding.sendYawn.setBounceClickListener { sendEmoji("🥱") }
        binding.chatBoxFriendly.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_SEND) {
                viewModel.sendMessage2FriendlyChat(binding.chatBoxFriendly.text.toString())?.also {
                    binding.chatBoxFriendly.setText("")
                } ?: toast("Write Something..")
                true // Return true to indicate that you have consumed the event
            } else {
                false // Return false to allow the system to handle the event
            }
        }
    }

    private fun sendEmoji(emoji: String) {
        viewModel.sendMessage2FriendlyChat(text = emoji)
        viewModel.ignoreDrawerClosesSound = true
        drawerLayout.closeDrawer(GravityCompat.START)
    }

    fun emojiRunner(gif: Int, sound: Int) {
        binding.sendHaha.isEnabled = false
        binding.sendCry.isEnabled = false
        binding.sendKiss.isEnabled = false
        binding.sendScream.isEnabled = false
        binding.sendYawn.isEnabled = false
        if (!pref.read("muted", false)) {
            val mediaPlayer = MediaPlayer.create(parentActivity, sound)
            mediaPlayer.start()
            mediaPlayer.setOnCompletionListener(MediaPlayer::release)
        }
        parentActivity.findViewById<ImageView>(R.id.emojiPlay).apply {
            loadDrawable(gif)
            show()
        }
        viewModel.setNewMsgBoltVisible(false)
        lifecycleScope.launch {
            delay(2500)
            parentActivity.findViewById<View>(R.id.emojiPlay).gone()
            binding.sendHaha.isEnabled = true
            binding.sendCry.isEnabled = true
            binding.sendKiss.isEnabled = true
            binding.sendScream.isEnabled = true
            binding.sendYawn.isEnabled = true
        }
    }


    companion object {
        var lastMsgKey = ""
    }
}